### DISCUSSION TITLE
Using event triggers and logging

### DISCUSSION THREAD

**Discussion by imreburn** [bitcraze/discussions#1047] [Q&A] [answered]
Hi,

I am trying to understand how event triggers are used with logging. I would use the event `estTDOA` in [estimator.c](https://github.com/bitcraze/crazyflie-firmware/blob/dbb9df1137f11d4e7e3771c56d25a7137b5b69cc/src/modules/src/estimator/estimator.c) and [config_kalman.txt](https://github.com/bitcraze/crazyflie-firmware/blob/dbb9df1137f11d4e7e3771c56d25a7137b5b69cc/tools/usdlog/config_kalman.txt) for example. 

1. `on:estTDOA` in config file, does it log whether this event is triggered?
2. There are payloads: `idA`, `idB`, `distanceDiff`, and how are they used here? When `on:estTDOA` is triggered and logged, these payloads are logged together automatically?

P.S. This is somewhat different, but can you explain the difference between three `modes` in logging? 

Thanks for your time!

**Comment by krichardsson** [comment, marked as answer]
There is [a bit of documentation](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/eventtrigger/)

It was a some time ago that I used the SD card logging and our flight lab is currently not usable so I'm doing a bit of guesswork.
@whoenig Feel free to clarify/correct ;-)

> on:estTDOA in config file, does it log whether this event is triggered?

Yes, you will log the toda data that is sent to the estimator.

If you plan to work on the kalman estimator, you also need acceleration and gyro data. An example config could be (I think...)
```
1     # version
2048  # buffer size in bytes
log   # file name
1     # enable on startup (0/1)
on:estTDOA
on:estGyroscope
gyro.x
gyro.y
gyro.z
on:estAcceleration
acc.x
acc.y
acc.z
```

> There are payloads: idA, idB, distanceDiff, and how are they used here? When on:estTDOA is triggered and logged, these payloads are logged together automatically?

I think they are logged if you add `on:estTDOA` to the config file.
There is also some python test code that might be useful to understand how to use the loged data. There is a pull request that was merged some time ago, these files are a good starting point I think. I you want to play with the kalman estimator, this is the way to go.

> P.S. This is somewhat different, but can you explain the difference between three modes in logging?

Do you mean the TWR, TDoA2 and TDoA3 modes?
For TWR you should use `on:estDistance`. From an estimator point of view TDoA2 and 3 are identical and you should use `on:estTDOA`

**Reply by whoenig** [reply]
Seems all correct! Some events have associated payloads, which will be automatically logged. In addition, one can add a list of "custom" log variables that will be stored at the time of the occurrence of the event. Each event also automatically has a timestamp (from the usec timer).

**Reply by imreburn** [reply]
Thank you for the answer! I used this example of the Kalman estimator to understand the event triggers. I am building a custom logging for my decentralized flight as an sd card is the only option for logging.

> P.S. This is somewhat different, but can you explain the difference between three modes in logging?
> Do you mean the TWR, TDoA2 and TDoA3 modes?

I found there are three options for the mode(`(0: disabled, 1: synchronous stabilizer, 2: asynchronous)`) in `config.txt` file, and I am wondering the differences between them.

**Reply by whoenig** [reply]
It's for the fixed frequency logging. 0 - disabled; 1 - log synchronously as part of the main loop; 2 - log in an independent task. 1 is more accurate as there should be no data races, but it could in theory have an impact on flight performance.

**Reply by imreburn** [reply]
Thanks for the answers. If the fixed frequency is not used, it seems that logging goes along with the main loop, whose frequency is not fixed. Is it right?

**Reply by whoenig** [reply]
Not the "main loop" (stabilizier loop), but it's triggered asynchronously by the events. This can happen from any FreeRTOS task.

**Reply by imreburn** [reply]
Thanks for the answer. I put my question incorrectly. I meant that the events in the `config_kalman.txt` could be triggered from each cycle of estimator, so I thought event logging here can be up to the frequency of the estimator loop. Thanks again for the explanation.

**Reply by whoenig** [reply]
Even higher - the events are triggered when measurements are pushed to the queue. For example you can log gyro information at 1kHz using the events.
